### PR TITLE
TYP: Typing hints in pandas/io/formats/{css,csvs}.py

### PR DESCRIPTION
**Pull request by ShaharNaveh**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/io/formats/css.py [hunk 1]
```diff
@@ -1,24 +1,39 @@
...
+        prop
+        value : str
+        """
         tokens = value.split()
         try:
             mapping = self.SIDE_SHORTHANDS[len(tokens)]
         except KeyError:
             warnings.warn(
-                f'Could not expand "{prop}: {value}"', CSSWarning,
+                f"Could not expand '{prop}: {value}'", CSSWarning,
```
Comment (WillAyd) [inline pandas/io/formats/css.py]: Shouldn't need to change this; the intention was for double quotes within the error message

File: pandas/io/formats/css.py [hunk 2]
```diff
@@ -172,7 +187,9 @@ def __call__(self, declarations_str, inherited=None):
 
     def size_to_pt(self, in_val, em_pt=None, conversions=UNIT_RATIOS):
         def _error():
-            warnings.warn(f"Unhandled size: {repr(in_val)}", CSSWarning)
+            warnings.warn(
+                f"Unhandled size: {repr(in_val)}", CSSWarning,
```
Comment (WillAyd) [inline pandas/io/formats/css.py]:
```suggestion
                f"Unhandled size: {in_val}", CSSWarning,
```

Shouldn't need the `repr` here any more

File: pandas/io/formats/csvs.py [hunk 3]
```diff
@@ -30,30 +32,54 @@
 
 
 class CSVFormatter:
+    r"""
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]: I think can delete this docstring; it doesn't add any value yet over what the annotations already tell you

File: pandas/io/formats/csvs.py [hunk 4]
```diff
@@ -324,8 +351,13 @@ def _save(self):
 
             self._save_chunk(start_i, end_i)
 
-    def _save_chunk(self, start_i: int, end_i: int):
-
+    def _save_chunk(self, start_i: int, end_i: int) -> None:
+        """
+        Parameters
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]: Same here; I think shouldn't need to add docstrings if they only repeat the annotations

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/formats/css.py [hunk 5]
```diff
@@ -1,24 +1,39 @@
...
+    This CSS syntax cannot currently be parsed.
+    """
 
     pass
 
 
 def _side_expander(prop_fmt: str):
-    def expand(self, prop, value):
+    """
+    Parameters
```
Comment (WillAyd) [inline pandas/io/formats/css.py]: Similar comment as before; not worth adding docstrings if they just repeat the annotations and nothing more

File: pandas/io/formats/css.py [hunk 6]
```diff
@@ -1,24 +1,39 @@
...
+        prop
+        value : str
+        """
         tokens = value.split()
         try:
             mapping = self.SIDE_SHORTHANDS[len(tokens)]
         except KeyError:
             warnings.warn(
-                f'Could not expand "{prop}: {value}"', CSSWarning,
+                f"Could not expand {prop}: {value}", CSSWarning,
```
Comment (WillAyd) [inline pandas/io/formats/css.py]: This isn't equivalent to the original exception message; I think saw on another PR - just need to be mindful of quotes

File: pandas/io/formats/css.py (hunk 2 above)
Comment (WillAyd) [reply pandas/io/formats/css.py]: I see marked as resolved but I think this is open; if you can commit changes or comment as to why not needed would be helpful

File: pandas/io/formats/csvs.py [hunk 7]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
     def __init__(
         self,
         obj,
-        path_or_buf=None,
-        sep=",",
-        na_rep="",
+        path_or_buf: Optional[FilePathOrBuffer[str]] = None,
+        sep: str = ",",
+        na_rep: str = "",
         float_format=None,
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        float_format: Optional[str] = None,
```

File: pandas/io/formats/csvs.py [hunk 8]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
...
         cols=None,
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        cols: Sequence[Hashable] = None,
```

File: pandas/io/formats/csvs.py [hunk 9]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
...
-        header=True,
-        index=True,
+        header: Union[bool, Sequence[str]] = True,
+        index: bool = True,
         index_label=None,
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        index_label: Optional[Union[bool, Hashable, Sequence[Hashable]]] = None,
```

File: pandas/io/formats/csvs.py [hunk 10]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
...
-        mode="w",
-        encoding=None,
+        mode: str = "w",
+        encoding: Optional[str] = None,
         compression="infer",
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        compression: str = "infer",
```

File: pandas/io/formats/csvs.py [hunk 11]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
...
-        quotechar='"',
-        date_format=None,
-        doublequote=True,
+        quoting: Optional[int] = None,
+        line_terminator: Optional[str] = "\n",
+        chunksize: Optional[int] = None,
+        quotechar: Optional[str] = '"',
+        date_format: Optional[str] = None,
+        doublequote: bool = True,
         escapechar=None,
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        escapechar: Optional[str] = None,
```

File: pandas/io/formats/csvs.py [hunk 12]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
...
-        decimal=".",
+        decimal: Optional[str] = ".",
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        decimal: str = ".",
```

I don't think this is optional

File: pandas/io/formats/csvs.py [hunk 13]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
...
-        quoting=None,
-        line_terminator="\n",
-        chunksize=None,
-        quotechar='"',
-        date_format=None,
-        doublequote=True,
+        quoting: Optional[int] = None,
+        line_terminator: Optional[str] = "\n",
+        chunksize: Optional[int] = None,
+        quotechar: Optional[str] = '"',
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        quotechar: str = '"',
```

I don't think this is optional

File: pandas/io/formats/csvs.py [hunk 14]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
...
+        encoding: Optional[str] = None,
         compression="infer",
-        quoting=None,
-        line_terminator="\n",
-        chunksize=None,
-        quotechar='"',
-        date_format=None,
-        doublequote=True,
+        quoting: Optional[int] = None,
+        line_terminator: Optional[str] = "\n",
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        line_terminator: str = "\n",
```

Don't think optional

File: pandas/io/formats/csvs.py [hunk 15]
```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
...
-        sep=",",
-        na_rep="",
+        path_or_buf: Optional[FilePathOrBuffer[str]] = None,
+        sep: str = ",",
+        na_rep: str = "",
         float_format=None,
         cols=None,
-        header=True,
-        index=True,
+        header: Union[bool, Sequence[str]] = True,
```
Comment (WillAyd) [inline pandas/io/formats/csvs.py]:
```suggestion
        header: Union[bool, Sequence[Hashable]] = True,
```

Yea the docstring says `str` but we really allow almost any hashable object as a column label, so just be mindful of that for annotations

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/formats/csvs.py (hunk 8 above)
Comment (MomIsBestFriend) [reply pandas/io/formats/csvs.py]: mypy's log for 3 diffrent annotations:


```cols: Sequence[Hashable] = None```

```
$ mypy pandas
pandas/io/formats/csvs.py:34: error: Incompatible default for argument "cols" (default has type "None", argument has type "Sequence[Hashable]")
pandas/core/generic.py:3171: error: Argument "cols" to "CSVFormatter" has incompatible type "Optional[Sequence[Optional[Hashable]]]"; expected "Sequence[Hashable]"
Found 2 errors in 2 files (checked 816 source files)
```
---

```cols: Optional[Sequence[Hashable]] = None```

```
$ mypy pandas
pandas/io/formats/csvs.py:128: error: Argument 1 to "list" has incompatible type "Optional[Sequence[Hashable]]"; expected "Iterable[Hashable]"
pandas/io/formats/csvs.py:139: error: Argument 1 to "len" has incompatible type "Optional[Sequence[Hashable]]"; expected "Sized"
pandas/core/generic.py:3171: error: Argument "cols" to "CSVFormatter" has incompatible type "Optional[Sequence[Optional[Hashable]]]"; expected "Optional[Sequence[Hashable]]"
Found 3 errors in 2 files (checked 816 source files)
```
---

```cols: Optional[Sequence[Optional[Hashable]]] = None```

```
$ mypy pandas
pandas/io/formats/csvs.py:128: error: Argument 1 to "list" has incompatible type "Optional[Sequence[Optional[Hashable]]]"; expected "Iterable[Optional[Hashable]]"
pandas/io/formats/csvs.py:139: error: Argument 1 to "len" has incompatible type "Optional[Sequence[Optional[Hashable]]]"; expected "Sized"
Found 2 errors in 2 files (checked 816 source files)
```

---

Any suggestions?

File: pandas/io/formats/csvs.py (hunk 10 above)
Comment (MomIsBestFriend) [reply pandas/io/formats/csvs.py]: Should it be typed as
```python
Union[str, Mapping[str, str], None]
```

?

---

mypy's log for 2 diffrent annotations:

```python
compression: str = "infer"
```

```
$ mypy pandas
pandas/io/formats/csvs.py:63: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")
pandas/core/generic.py:3167: error: Argument "compression" to "CSVFormatter" has incompatible type "Union[str, Mapping[str, str], None]"; expected "str"
Found 2 errors in 2 files (checked 816 source files)
```

---

```python
compression: Union[str, Mapping[str, str], None] = "infer"
```

```
$ mypy pandas
Success: no issues found in 816 source files
```

File: pandas/io/formats/csvs.py (hunk 12 above)
Comment (MomIsBestFriend) [reply pandas/io/formats/csvs.py]: mypy raises error, saying it's expecting ```Optional[str]```

File: pandas/io/formats/csvs.py (hunk 13 above)
Comment (MomIsBestFriend) [reply pandas/io/formats/csvs.py]: Keep it as 

```python
quotechar: Optional[str] = '"'
```
?

---

mypy's log:

```
$ mypy pandas
pandas/io/formats/csvs.py:88: error: Incompatible types in assignment (expression has type "None", variable has type "str")
Found 1 error in 1 file (checked 816 source files)
```

File: pandas/io/formats/csvs.py (hunk 14 above)
Comment (MomIsBestFriend) [reply pandas/io/formats/csvs.py]: Should I keep it as

```python
line_terminator: Optional[str] = "\n"
```

?

---

mypy's log:

```
$ mypy pandas
pandas/core/generic.py:3164: error: Argument "line_terminator" to "CSVFormatter" has incompatible type "Optional[str]"; expected "str"
Found 1 error in 1 file (checked 816 source files)
```

File: pandas/io/formats/csvs.py (hunk 8 above)
Comment (WillAyd) [reply pandas/io/formats/csvs.py]: It looks like the annotations for `columns` might not be consistent in the IO methods. Can you open up a follow up issue for this and just leave unannotated for now? I think might be more difficult so OK to branch off given there's already a good deal in this PR

File: pandas/io/formats/csvs.py (hunk 10 above)
Comment (WillAyd) [reply pandas/io/formats/csvs.py]: Hmm OK I guess should match as `compression: Optional[Union[str, Mapping[str, str]]] = "infer",`

File: pandas/io/formats/csvs.py (hunk 14 above)
Comment (WillAyd) [reply pandas/io/formats/csvs.py]: Hmm there is a minor disconnect between `to_csv` and the `CSVFormatter` class on this. Can you create a follow up issue? Not sure of easiest way to correct

File: pandas/io/formats/csvs.py (hunk 13 above)
Comment (WillAyd) [reply pandas/io/formats/csvs.py]: I think similar comment as line_terminator

File: pandas/io/formats/csvs.py (hunk 12 above)
Comment (WillAyd) [reply pandas/io/formats/csvs.py]: Same comment

**Review by WillAyd — COMMENTED** [review]
OK for any of the errors we just have a disconnect between defaults for CSVFormatter and to_csv. These may or may not necessarily be issues but just separate out and if you can open up a follow up issue to address would be ideal

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
@simonjayhawkins care to look?

File: pandas/io/formats/csvs.py [hunk 16]
```diff
@@ -33,27 +34,26 @@ class CSVFormatter:
...
-        mode="w",
-        encoding=None,
-        compression="infer",
-        quoting=None,
+        header: Union[bool, Sequence[Hashable]] = True,
+        index: bool = True,
+        index_label: Optional[Union[bool, Hashable, Sequence[Hashable]]] = None,
+        mode: str = "w",
+        encoding: Optional[str] = None,
+        compression: Union[str, Mapping[str, str], None] = "infer",
```
Comment (simonjayhawkins) [inline pandas/io/formats/csvs.py]:
```suggestion
        compression: Optional[Union[str, Mapping[str, str]]] = "infer",
```

File: pandas/io/formats/csvs.py [hunk 17]
```diff
@@ -33,27 +34,26 @@ class CSVFormatter:
...
+        float_format: Optional[str] = None,
         cols=None,
-        header=True,
-        index=True,
-        index_label=None,
-        mode="w",
-        encoding=None,
-        compression="infer",
-        quoting=None,
+        header: Union[bool, Sequence[Hashable]] = True,
```
Comment (simonjayhawkins) [inline pandas/io/formats/csvs.py]: question for @WillAyd - do we want ` Sequence[Optional[Hashable]]] ` or is this for readability and only add `Optional` when mypy complains?

Comment (WillAyd) [reply pandas/io/formats/csvs.py]: Personal preference for only when complaining but it isn't strong if you feel otherwise

Comment (simonjayhawkins) [reply pandas/io/formats/csvs.py]: fine with that. (although may need to make exception for public apis)

**Comment by jreback** [comment]
thanks @MomIsBestFriend

**Pull request merged**